packages/transport-native-ble/src/api/nativeBleManager.ts

```diff
@@ -0,0 +1,438 @@
+            await device.characteristicsForService(NUS_SERVICE_UUID);
+
+        // debugLog('Characteristics: ', JSON.stringify(characteristics));
+
+        if (characteristics.length === 0) {
+            throw new Error(
+                'No device characteristics found. Make sure the device is connected and has the correct service UUID.',
+            );
+        }
+
+        let writeCharacteristic: Characteristic | undefined;
+        let notifyCharacteristic: Characteristic | undefined;
+        for (const characteristic of characteristics) {
+            if (characteristic.isWritableWithoutResponse) {
+                debugLog('Found write characteristic: ', characteristic.uuid);
+                writeCharacteristic = characteristic;
+            } else if (characteristic.isNotifiable) {
+                debugLog('Found notify characteristic: ', characteristic.uuid);
+                notifyCharacteristic = characteristic;
+            } else {
+                debugLog('Found other unknown characteristic: ', characteristic.uuid);
+            }
+        }
+
+        if (!writeCharacteristic) {
+            throw new Error('Write characteristic not found.');
+        }
+
+        if (!notifyCharacteristic) {
+            throw new Error('Notify characteristic not found.');
+        }
+
+        device.monitorCharacteristicForService(
+            NUS_SERVICE_UUID,
+            notifyCharacteristic.uuid,
+            (error, characteristic) => {
+                if (error) {
+                    console.error('Error monitoring characteristic: ', error);
+
+                    return;
+                }
+                if (characteristic) {
+                    debugLog('Received data: ', characteristic.value);
+                    if (characteristic?.value) {
+                        this.addDeviceReadOutput(device.id, characteristic.value);
+                    }
+                } else {
+                    console.error('No characteristic received');
+                }
+            },
+        );
+
+        this.addDeviceToAppConnectedDevices({
+            bleDevice: device,
+            writeUuid: writeCharacteristic.uuid,
+            notifyUuid: notifyCharacteristic.uuid,
+            readOutputQueue: [],
+        });
+
+        debugLog(`Device connected: ${device.id}`);
+        debugLog(`Device manufacturerData: ${device.manufacturerData}`);
+
+        this.bleManager.onDeviceDisconnected(device.id, async (error, disconnectedDevice) => {
+            if (error) {
+                console.error('Device disconnected error', error);
+            }
+            if (disconnectedDevice) {
+                this.removeDeviceFromAppConnectedDevices(disconnectedDevice.id);
+            }
+        });
+
+        return device;
+    };
+
+    public getAllConnectedDevices = async () => {
+        return this.appConnectedDevices.map(d => d.bleDevice);
+    };
+
+    public findDevice = (deviceId: string) => {
```

Comment:
nit for better clarity:

```suggestion
    public findConnectedDevice = (deviceId: string) => {
```